File: src/main/kotlin/com/foolchen/kotlin/java/object/DelegateExample.java
package com.foolchen.kotlin.java.object;

/**
 * @author chenchong. Created on 2017/6/4.
 */
public class DelegateExample {
  interface Base {
    void print();
  }

  static class BaseImpl implements Base {
    int x;

    public BaseImpl(int x) {
      this.x = x;
    }

    @Override
    public void print() {
      System.out.println(x);
    }
  }

  static class Derived implements Base {
    Base base;

    public Derived(Base base) {
      this.base = base;
    }

    @Override
    public void print() {
      this.base.print();
    }
  }

  public static void main(String[] args) {
    new Derived(new BaseImpl(10)).print();
  }
}

File: src/main/kotlin/com/foolchen/kotlin/java/object/DelegateExample.kt
package com.foolchen.kotlin.java.`object`

/**
 * 代理模式示例,此处的演示可以与Java中的实现进行比较[DelegateExample]
 * @sample DelegateExample
 * @author chenchong. Created on 2017/6/4.
 */

interface Base {
    fun print()
}

class BaseImpl(val x: Int) : Base {
    override fun print() {
        println(x)
    }
}

/*
“Derived 的超类型列表中的 by-子句表示 b 将会在 Derived 中内部存储。
并且编译器将生成转发给 b 的所有 Base 的方法。”
*/
class Derived(b: Base) : Base by b

fun main(args: Array<String>) {
    val b = BaseImpl(10)
    Derived(b).print()
}